File: packages/engine/stdlib/src/py/hstd/neighbor.py
"""
Neighbor utility functions.
"""
from typing import List

from .spatial import distance_between
from .agent import AgentFieldError, AgentState


def neighbors_on_position(agent: AgentState, neighbors: List[AgentState]) -> List[AgentState]:
    """
    Returns all `neighbors` whose position is identical to the `agent`.
    """
    if agent.position is None:
        raise AgentFieldError(agent.agent_id, "position", "cannot be None")

    return [n for n in neighbors if n.position == agent.position]


def neighbors_in_radius(
    agent: AgentState,
    neighbors: List[AgentState],
    max_radius: float = 1,
    min_radius: float = 0,
    distance_function: str = "euclidean",
    z_axis: bool = False,
) -> List[AgentState]:
    """
    Returns all neighbors within a certain vision radius of an agent.
    Default is 2D (`z_axis` set to false). Set `z_axis` to true for 3D positions.

    Args:
        agent: central agent
        neighbors: context.neighbors() array, or an array of agents
        max_radius: minimum radius for valid neighbors
        min_radius: maximum radius for valid neighbors
        distance_function: type of distance function to use
        z_axis: include z-axis in distance calculations
    """

    if agent.position is None:
        raise AgentFieldError(agent.agent_id, "position", "cannot be None")

    def in_radius(neighbor: AgentState) -> bool:
        if neighbor.position is None:
            return False

        d = distance_between(neighbor, agent, distance_function, z_axis)
        if d is None:
            return False

        return (d <= max_radius) and (d >= min_radius)

    return [n for n in neighbors if in_radius(n)]


def difference_vector(vec1: List[float], vec2: List[float]):
    """
    Calculate the difference vector `vec2` - `vec1`.
    """
    return [vec2[ind] - vec1[ind] for ind in range(len(vec1))]


def in_front_planar(agent: AgentState, neighbor: AgentState) -> bool:
    """
    Return True if a neighbor is anywhere in front of the agent.
    """

    a_dir = agent["direction"]

    [dx, dy, dz] = difference_vector(agent["position"], neighbor["position"])
    D = a_dir[0] * dx + a_dir[1] * dy + a_dir[2] * dz

    return D > 0


def is_linear(agent: AgentState, neighbor: AgentState, front: bool) -> bool:
    """
    Check if a neighbor lies along the direction vector of the agent, and is
    in front of or behind the agent, based on `front`.
    """
    [dx, dy, dz] = difference_vector(agent["position"], neighbor["position"])
    [ax, ay, az] = agent["direction"]

    cross_product = [dy * az - dz * ay, dx * az - dz * ax, dx * ay - dy * ax]

    if cross_product != [0, 0, 0]:
        return False

    # check if same direction
    same_dir = (ax * dx > 0) or (ay * dy > 0) or (az * dz > 0)

    return same_dir is front


def neighbors_in_front(
    agent: AgentState, neighbors: List[AgentState], colinear: bool = False
) -> List[AgentState]:
    """
    Return all `neighbors` in front of the `agent`. If `colinear` is True
    check that the neighbor lies along the agent's direction vector.
    """

    if agent.position is None:
        raise AgentFieldError(agent.agent_id, "position", "cannot be None")
    if agent.direction is None:
        raise AgentFieldError(agent.agent_id, "direction", "cannot be None")

    if colinear:
        return [n for n in neighbors if is_linear(agent, n, True)]
    else:
        return [n for n in neighbors if in_front_planar(agent, n)]


def neighbors_behind(
    agent: AgentState, neighbors: List[AgentState], colinear: bool = False
) -> List[AgentState]:
    """
    Return all `neighbors` behind the `agent`. If `colinear` is True
    check that the neighbor lies along the agent's direction vector.
    """

    if agent.position is None:
        raise AgentFieldError(agent.agent_id, "position", "cannot be None")
    if agent.direction is None:
        raise AgentFieldError(agent.agent_id, "direction", "cannot be None")

    if colinear:
        return [n for n in neighbors if is_linear(agent, n, False)]
    else:
        return [n for n in neighbors if not in_front_planar(agent, n)]

File: apps/engine/stdlib/src/ts/neighbor.ts
/** Neighbor Functions */
import { PotentialAgent } from "./agent";
import { Distance, distanceBetween } from "./spatial";

const posError = new Error("agent must have a position");
const dirError = new Error("agent must have a direction");

/**
 * Returns all neighbors that share an agent's position
 * @param agent
 * @param neighbors - context.neighbors() array, or an array of agents
 * */
export function neighborsOnPosition(
  agent: PotentialAgent,
  neighbors: PotentialAgent[],
) {
  return neighbors.filter((neighbor) => {
    const aPos = agent.position;
    const nPos = neighbor.position;

    if (!aPos || !nPos) {
      throw posError;
    }

    for (let i = 0; i < aPos.length; i++) {
      if (aPos[i] !== nPos[i]) {
        return false;
      }
    }

    return true;
  });
}

/**
 * Returns all neighbors within a certain vision radius of an agent.
 * Default is 2D (`z_axis` set to false). Set `z_axis` to true for 3D positions.
 * @param agent
 * @param neighbors - context.neighbors() array, or an array of agents
 * @param max_radius - defaults to 1
 * @param min_radius - defaults to 0
 * @param distanceFunction - defaults to "euclidean"
 * @param z_axis - defaults to false
 */
export function neighborsInRadius(
  agent: PotentialAgent,
  neighbors: PotentialAgent[],
  max_radius: number = 1,
  min_radius: number = 0,
  distanceFunction: Distance = "euclidean",
  z_axis: boolean = false,
) {
  const aPos = agent.position;
  if (!aPos) {
    throw posError;
  }

  return neighbors.filter((neighbor) => {
    const nPos = neighbor.position;
    if (!nPos) {
      return false;
    }

    const d = distanceBetween(neighbor, agent, distanceFunction, z_axis);
    return d <= max_radius && d >= min_radius;
  });
}

/**
 * Searches and returns all neighbors whose positions are in front of an agent.
 * Default is set to planar calculations and will return all neighbors located
 * in front of the plane created by the agent's direction
 * @param agent
 * @param neighbors - context.neighbors() array or array of agents
 * @param colinear - If set to true, will return all agents on the same line as the agent.
 */
export function neighborsInFront(
  agent: PotentialAgent,
  neighbors: PotentialAgent[],
  colinear: boolean = false,
) {
  return neighbors.filter((neighbor) => {
    const aPos = agent.position;
    const aDir = agent.direction;
    const nPos = neighbor.position;

    if (!aPos || !nPos) {
      throw posError;
    }
    if (!aDir) {
      throw dirError;
    }

    if (colinear) {
      const count = aDir.reduce(function (n, val) {
        return n + val;
      }, 0);

      const dx = nPos[0] - aPos[0];
      const dy = nPos[1] - aPos[1];
      const dz = nPos[2] - aPos[2];

      const xt = aDir[0] !== 0 ? dx / aDir[0] : 0;
      const yt = aDir[1] !== 0 ? dy / aDir[1] : 0;
      const zt = aDir[2] !== 0 ? dz / aDir[2] : 0;

      // Only one direction set - only that axis value can change
      if (count === 1) {
        return (
          (xt > 0 && dy === 0 && dz === 0) ||
          (yt > 0 && dx === 0 && dz === 0) ||
          (zt > 0 && dx === 0 && dy === 0)
        );
      } else if (count === 2) {
        // Two directions set
        return (
          (xt === yt && xt > 0 && dz === 0) ||
          (yt === zt && yt > 0 && dx === 0) ||
          (xt === zt && xt > 0 && dy === 0)
        );
      } else if (count === 3) {
        // Three directions set
        return xt === yt && yt === zt && xt > 0;
      }

      // Else direction is [0,0,0] or an unsupported direction array
      return false;
    }

    // Planar calculations
    const dx = nPos[0] - aPos[0];
    const dy = nPos[1] - aPos[1];
    const dz = nPos[2] - aPos[2];

    const D = aDir[0] * dx + aDir[1] * dy + aDir[2] * dz;

    if (D <= 0) {
      return false;
    }

    return true;
  });
}

/**
 * Searches and returns all neighbors whose positions are behind an agent.
 * Default is set to planar calculations and will return all neighbors located
 * behind the plane created by the agent's direction
 * @param agent
 * @param neighbors - context.neighbors() array or array of agents
 * @param colinear - If set to true, will return all neighbors on the same line as the
 * agent a. Defaults to false
 */
export function neighborsBehind(
  agent: PotentialAgent,
  neighbors: PotentialAgent[],
  colinear = false,
) {
  return neighbors.filter((neighbor) => {
    const aPos = agent.position;
    const aDir = agent.direction;
    const nPos = neighbor.position;

    if (!aPos || !nPos) {
      throw posError;
    }
    if (!aDir) {
      throw dirError;
    }

    if (colinear) {
      const count = aDir.reduce(function (n, val) {
        return n + val;
      }, 0);

      const dx = nPos[0] - aPos[0];
      const dy = nPos[1] - aPos[1];
      const dz = nPos[2] - aPos[2];

      const xt = aDir[0] !== 0 ? (nPos[0] - aPos[0]) / aDir[0] : 0;
      const yt = aDir[1] !== 0 ? (nPos[1] - aPos[1]) / aDir[1] : 0;
      const zt = aDir[2] !== 0 ? (nPos[2] - aPos[2]) / aDir[2] : 0;

      // Only one direction set - only that axis value can change
      if (count === 1) {
        return (
          (xt < 0 && dy === 0 && dz === 0) ||
          (yt < 0 && dx === 0 && dz === 0) ||
          (zt < 0 && dx === 0 && dy === 0)
        );
      } else if (count === 2) {
        // Two directions set
        return (
          (xt === yt && xt < 0 && dz === 0) ||
          (yt === zt && yt < 0 && dx === 0) ||
          (xt === zt && xt < 0 && dy === 0)
        );
      } else if (count === 3) {
        // Three directions set
        return xt === yt && yt === zt && xt < 0;
      }

      // Else direction is [0,0,0] or an unsupported direction array
      return false;
    }

    // Planar calculations
    const dx = nPos[0] - aPos[0];
    const dy = nPos[1] - aPos[1];
    const dz = nPos[2] - aPos[2];

    const D = aDir[0] * dx + aDir[1] * dy + aDir[2] * dz;

    if (D >= 0) {
      return false;
    }

    return true;
  });
}
